builtin/credential/userpass/path_users.go

```diff
@@ -95,43 +120,59 @@ func (b *backend) pathUserRead(
 	}, nil
 }
 
-func (b *backend) pathUserWrite(
-	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-	name := strings.ToLower(d.Get("name").(string))
-	password := d.Get("password").(string)
-	policies := strings.Split(d.Get("policies").(string), ",")
-	for i, p := range policies {
-		policies[i] = strings.TrimSpace(p)
+func (b *backend) userCreateUpdate(req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	username := strings.ToLower(d.Get("username").(string))
+	userEntry, err := b.user(req.Storage, username)
+	if err != nil {
+		return nil, err
+	}
+	// Due to existence check, user will only be nil if it's a create operation
+	if userEntry == nil {
+		userEntry = &UserEntry{}
 	}
 
-	// Generate a hash of the password
-	hash, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+	// "password" will always be set here
+	err = b.updateUserPassword(req, d, userEntry)
 	if err != nil {
 		return nil, err
 	}
 
-	ttlStr := d.Get("ttl").(string)
-	maxTTLStr := d.Get("max_ttl").(string)
-	ttl, maxTTL, err := b.SanitizeTTL(ttlStr, maxTTLStr)
-	if err != nil {
-		return logical.ErrorResponse(fmt.Sprintf("err: %s", err)), nil
+	if _, ok := d.GetOk("policies"); ok {
+		err = b.updateUserPolicies(req, d, userEntry)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	// Store it
-	entry, err := logical.StorageEntryJSON("user/"+name, &UserEntry{
-		PasswordHash: hash,
-		Policies:     policies,
-		TTL:          ttl,
-		MaxTTL:       maxTTL,
-	})
-	if err != nil {
-		return nil, err
+	ttlStr := ""
```

Comment:
The problem here, and with `maxTTLStr`, is that if the value is already in the userEntry you wipe it out by using this empty string here. Set `ttlStr := userEntry.TTL.String()` and `maxTTLStr := userEntry.MaxTTL.String()`. That way, if it's not explicitly given (via the `GetOk`), and it's not a `CreateOperation`, the values passed into `SanitizeTTL` will be the string versions of the existing values without change, which will spit the parsed versions back out. The round trip isn't nice but it works.